File: backend/app/schemas/enabler.py
"""
Enabler Schemas

Key Enabler API 요청/응답 스키마
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, field_validator


class EnablerBase(BaseModel):
    """Enabler 기본 스키마"""

    name: str
    description: Optional[str] = None
    type: str
    planned_delivery_date: datetime
    actual_delivery_date: Optional[datetime] = None
    status: str = "requested"
    criticality: str = "medium"
    responsible_person: Optional[str] = None
    notes: Optional[str] = None

    @field_validator("type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        """
        타입 값 검증

        허용되는 타입: document, equipment, approval, resource, license, training
        """
        allowed_types = [
            "document",
            "equipment",
            "approval",
            "resource",
            "license",
            "training",
        ]
        if v not in allowed_types:
            raise ValueError(f"type must be one of {allowed_types}, got '{v}'")
        return v

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        """
        상태 값 검증

        허용되는 상태: requested, in_progress, delivered, delayed, cancelled
        """
        allowed_statuses = [
            "requested",
            "in_progress",
            "delivered",
            "delayed",
            "cancelled",
        ]
        if v not in allowed_statuses:
            raise ValueError(f"status must be one of {allowed_statuses}, got '{v}'")
        return v

    @field_validator("criticality")
    @classmethod
    def validate_criticality(cls, v: str) -> str:
        """
        중요도 값 검증

        허용되는 중요도: low, medium, high, critical
        """
        allowed_criticalities = ["low", "medium", "high", "critical"]
        if v not in allowed_criticalities:
            raise ValueError(f"criticality must be one of {allowed_criticalities}, got '{v}'")
        return v


class EnablerCreate(EnablerBase):
    """Enabler 생성 요청 스키마"""

    project_id: int


class EnablerUpdate(BaseModel):
    """Enabler 수정 요청 스키마"""

    name: Optional[str] = None
    description: Optional[str] = None
    type: Optional[str] = None
    planned_delivery_date: Optional[datetime] = None
    actual_delivery_date: Optional[datetime] = None
    status: Optional[str] = None
    criticality: Optional[str] = None
    responsible_person: Optional[str] = None
    notes: Optional[str] = None

    @field_validator("type")
    @classmethod
    def validate_type(cls, v: Optional[str]) -> Optional[str]:
        """타입 값 검증"""
        if v is None:
            return v
        allowed_types = [
            "document",
            "equipment",
            "approval",
            "resource",
            "license",
            "training",
        ]
        if v not in allowed_types:
            raise ValueError(f"type must be one of {allowed_types}, got '{v}'")
        return v

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        """상태 값 검증"""
        if v is None:
            return v
        allowed_statuses = [
            "requested",
            "in_progress",
            "delivered",
            "delayed",
            "cancelled",
        ]
        if v not in allowed_statuses:
            raise ValueError(f"status must be one of {allowed_statuses}, got '{v}'")
        return v

    @field_validator("criticality")
    @classmethod
    def validate_criticality(cls, v: Optional[str]) -> Optional[str]:
        """중요도 값 검증"""
        if v is None:
            return v
        allowed_criticalities = ["low", "medium", "high", "critical"]
        if v not in allowed_criticalities:
            raise ValueError(f"criticality must be one of {allowed_criticalities}, got '{v}'")
        return v


class EnablerResponse(EnablerBase):
    """Enabler 응답 스키마"""

    id: int
    project_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/enabler.ts
/**
 * Enabler Types
 *
 * Enabler(Key Enabler) 관련 타입 정의
 */

/**
 * Enabler Type
 */
export type EnablerType =
  | 'document'
  | 'equipment'
  | 'approval'
  | 'resource'
  | 'license'
  | 'training';

/**
 * Enabler Status
 */
export type EnablerStatus =
  | 'requested'
  | 'in_progress'
  | 'delivered'
  | 'delayed'
  | 'cancelled';

/**
 * Enabler Criticality
 */
export type EnablerCriticality = 'low' | 'medium' | 'high' | 'critical';

/**
 * Enabler 인터페이스
 */
export interface Enabler {
  id: number;
  project_id: number;
  name: string;
  description?: string;
  type: EnablerType;
  planned_delivery_date: string;
  actual_delivery_date?: string;
  status: EnablerStatus;
  criticality: EnablerCriticality;
  responsible_person?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

/**
 * Enabler 생성 요청 인터페이스
 */
export interface EnablerCreate {
  project_id: number;
  name: string;
  description?: string;
  type: EnablerType;
  planned_delivery_date: string;
  actual_delivery_date?: string;
  status?: EnablerStatus;
  criticality?: EnablerCriticality;
  responsible_person?: string;
  notes?: string;
}

/**
 * Enabler 수정 요청 인터페이스
 */
export interface EnablerUpdate {
  name?: string;
  description?: string;
  type?: EnablerType;
  planned_delivery_date?: string;
  actual_delivery_date?: string;
  status?: EnablerStatus;
  criticality?: EnablerCriticality;
  responsible_person?: string;
  notes?: string;
}

/**
 * Enabler 타입 레이블
 */
export const EnablerTypeLabels: Record<EnablerType, string> = {
  document: '문서',
  equipment: '장비',
  approval: '승인',
  resource: '리소스',
  license: '라이선스',
  training: '교육',
};

/**
 * Enabler 상태 레이블
 */
export const EnablerStatusLabels: Record<EnablerStatus, string> = {
  requested: '요청됨',
  in_progress: '진행 중',
  delivered: '전달 완료',
  delayed: '지연됨',
  cancelled: '취소됨',
};

/**
 * Enabler 상태 색상
 */
export const EnablerStatusColors: Record<EnablerStatus, string> = {
  requested: '#9e9e9e',
  in_progress: '#2196f3',
  delivered: '#4caf50',
  delayed: '#ff9800',
  cancelled: '#f44336',
};

/**
 * Enabler 중요도 레이블
 */
export const EnablerCriticalityLabels: Record<EnablerCriticality, string> = {
  low: '낮음',
  medium: '보통',
  high: '높음',
  critical: '긴급',
};

/**
 * Enabler 중요도 색상
 */
export const EnablerCriticalityColors: Record<EnablerCriticality, string> = {
  low: '#4caf50',
  medium: '#ff9800',
  high: '#ff5722',
  critical: '#d32f2f',
};
